initial pass of resharing end to end

CluEleSsUK opened the pull request:
- updated to go 1.22
- presently newly added nodes don't work (only resharing the same group)
- more tests needed

CluEleSsUK committed:
added local DKG state for users

CluEleSsUK committed:
added reshare and state to sidecar

- created a state directory for sidecar
- refactored the key to be in relation to it
- implemented resharing for the same group

CluEleSsUK committed:
finished initial pass of resharing end to end

- updated to go 1.22
- presently newly added nodes don't work (only resharing the same group)
- more tests needed

CluEleSsUK committed:
linter

CluEleSsUK committed:
moved keys from localhost to 127 for github runner

AnomalRoil commented on cli/README.md:
````diff
@@ -42,8 +42,23 @@ $ ssv-dkg sign --input /path/to/deposit/data \
 ```
 Each operator must be in the form `$validatorNonce,$address`. Providing the wrong validator nonce may result in disaster for your DKG.
+The output directory will default to `~/.ssv`. It will be in a file named after the date (and a counter if you create multiple clusters in a day). 
+You will need to maintain this state file if you wish to reshare the key for this cluster in future, e.g. if operators become unresponsive and you wish to exclude them. 
````
```suggestion
You will need to maintain this state file if you wish to reshare the key for this cluster in the future, e.g. if operators become unresponsive and you wish to exclude them. 
```

AnomalRoil commented on cli/internal/cmd/reshare.go:
```diff
@@ -0,0 +1,86 @@
+	log.MaybeLog(fmt.Sprintf("✅ reshare completed successfully. Encrypted shares stored in %s", stateFilePath))
+}
+
+func operatorsOrStdin(cmd *cobra.Command) ([]string, error) {
+	if len(operatorFlag) != 0 {
+		return operatorFlag, nil
+	}
+	// if the operator flag isn't passed, we consume operator addresses from stdin
+	stdin, err := io.ReadAll(cmd.InOrStdin())
+	if err != nil {
+		return nil, errors.New("error reading from stdin")
+	}
+
+	operatorString := strings.Trim(string(stdin), "\n")
+	if operatorString == "" {
+		return nil, errors.New("you must provider either the --operator flag or operators via stdin")
+	}
+
+	return strings.Split(operatorString, " "), nil
+}
```
I feel this could have deserved a bit more separation in order to have some tests

CluEleSsUK replied:
have pulled it into its own function

CluEleSsUK committed:
resharing now works for groups with new participants

- user passes a hash of the old key share, to avoid confusion if some
  DKG fails for some nodes
- integration test added for multiple reshares and with new groups
- threshold isn't passed around, as we default to 50%+1
- added a debug logging flag
- added lots of logs

CluEleSsUK committed:
graceful shutdown of stub server to avoid port contention

AnomalRoil commented on cli/README.md:
````diff
@@ -42,8 +42,23 @@ $ ssv-dkg sign --input /path/to/deposit/data \
+⏳ starting distributed key resharing
+✅ distributed key reshared successfully!
+```
+Note: you will have to maintain 50% + 1 of the same operators in the new cluster for this to succeed.
````
Not mega clear, might be rephrased maybe

AnomalRoil commented on tools/stub/ssv-node-stub.go:
these stub stuff might belong in `internal` rather, no?

CluEleSsUK replied:
users may want to run the stub, so it's probably not `internal`

AnomalRoil commented on tools/stub/ssv-node-stub.go:
```diff
@@ -19,29 +22,52 @@ func StartStub(port uint) {
+	}()
+
+	return func() {
+		stop <- true
```
```suggestion
		close(stop)
```
so that it can be read as many time as you want in as many go routines as needed

AnomalRoil commented on cli/internal/cmd/reshare.go:
```diff
@@ -0,0 +1,86 @@
+		return nil, errors.New("error reading from stdin")
+	}
+
+	operatorString := strings.Trim(string(stdin), "\n")
```
`Trim` is only removing leading and trailing newspaces, is that just what you want here?

AnomalRoil commented on cli/internal/cmd/reshare.go:
```diff
@@ -0,0 +1,86 @@
+		return nil, errors.New("you must provider either the --operator flag or operators via stdin")
+	}
+
+	return strings.Split(operatorString, " "), nil
```
so you're expecting operators to be provided in stdin as:
```
1,operator 2,opop 3,perper 4,ratorrator
```
am I reading that code correctly?

AnomalRoil commented on cli/internal/cmd/sign.go:
```diff
@@ -19,7 +21,7 @@ var stateDirectory string
 	Use:   "sign",
 	Short: "Signs ETH deposit data by forming a validator cluster",
-	Long:  "Signs ETH deposit data by forming a validator cluster that creates a distributed key.",
+	Long:  "Signs ETH deposit data by forming a validator cluster that creates a distributed key. Operators can be passed via stdin.",
```
I'd specify the format we're expecting too.

CluEleSsUK replied:
Right now there isn't any validation on the format, though perhaps there ought to be

AnomalRoil commented on cli/internal/cmd/sign.go:
```diff
@@ -38,33 +40,60 @@ func init() {
+		}
 
-	log := shared.QuietLogger{Quiet: shortFlag}
+		args = strings.Split(operatorString, " ")
```
I really think we should have some function somewhere doing that since we need it in more than one place and it's tricky if we change it in one place and not in the other later on.

CluEleSsUK replied:
have refactored the other function to do both

AnomalRoil commented on cli/reshare.go:
```diff
@@ -0,0 +1,170 @@
+
+	// we do some sanity checks on the returned details to ensure the public key hasn't changed
+	err = verifyPublicPolynomialSameReshare(operatorResponses)
+	if err != nil {
```
```suggestion
	if err = verifyPublicPolynomialSameReshare(operatorResponses); err != nil {
```

AnomalRoil commented on integration/dkg_error_processes.go:
should probably be in `internal`

AnomalRoil commented on integration/integration_test.go:
should probably be in `internal` ?

AnomalRoil reviewed (changes requested):
First batch of comments

CluEleSsUK committed:
fixed some PR feedback, moved some things to internal
